File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.java
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        Map<Integer, ListNode> prefixDict = new HashMap<>();
        prefixDict.put(0, dummy);
        ListNode node = head;
        int prefix = 0;

        while (node != null) {
            prefix += node.val;
            if (prefixDict.containsKey(prefix)) {
                ListNode target = prefixDict.get(prefix).next;
                int tempSum = prefix + target.val;
                while (target != node) {
                    prefixDict.remove(tempSum);
                    target = target.next;
                    tempSum += target.val;
                }
                prefixDict.get(prefix).next = node.next;
            } else {
                prefixDict.put(prefix, node);
            }
            node = node.next;
        }

        return dummy.next;
    }
}

File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.py
class Solution:
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        prefix_dict = {0: dummy}
        node = head
        prefix = 0

        while node:
            prefix += node.val
            if prefix in prefix_dict:
                target = prefix_dict[prefix].next
                temp_sum = prefix + target.val
                while target != node:
                    del prefix_dict[temp_sum]
                    target = target.next
                    temp_sum += target.val
                prefix_dict[prefix].next = node.next
            else:
                prefix_dict[prefix] = node
            node = node.next

        return dummy.next